### PR TITLE
Finish Chapter13

### PR DESCRIPTION
**Pull request by heimrih**
Chapter 13 - microposts

### CODE REVIEW

**Comment by heimrih** [comment]
ready

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -15,4 +15,13 @@ def extract_locale
...
     {locale: I18n.locale}
   end
+
+  # Confirms a logged-in user.
+  def logged_in_user
+    return if logged_in?
+
+      store_location
+      flash[:danger] = "user.login"
+      redirect_to login_path
```
Comment (saobv-2223) [inline app/controllers/application_controller.rb]:
```suggestion
    store_location
    flash[:danger] = t "user.login"
    redirect_to login_path
```

File: app/helpers/microposts_helper.rb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+module MicropostsHelper
+end
```
Comment (saobv-2223) [inline app/helpers/microposts_helper.rb]: remove file that never used

File: app/views/shared/_micropost_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<%= form_for(@micropost) do |f| %>
```
Comment (saobv-2223) [inline app/views/shared/_micropost_form.html.erb]: remove ()

File: app/views/shared/_micropost_form.html.erb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+<% end %>
+
+<script type="text/javascript">
+  $("#micropost_image").bind("change", function() {
+    var size_in_megabytes = this.files[0].size/1024/1024;
+    if (size_in_megabytes > 5) {
+      alert( <%= t("micropost.error.alert")%> );
+    }
+  });
+</script>
```
Comment (saobv-2223) [inline app/views/shared/_micropost_form.html.erb]: Do not write js in html file, you should write to another js file and required to application.js

File: app/views/static_pages/home.html.erb [hunk 5]
```diff
@@ -1,13 +1,27 @@
...
-</h2>
-<%= link_to t("signup"), signup_path, class: "btn btn-lg btn-primary" %>
-</div>
-<%= link_to image_tag("rails-1-logo.png", alt: "Rails logo", width:
-"200px"),"https://rubyonrails.org/" %>
+<% if logged_in? %>
+  <div class="row">
+    <aside class="col-md-4">
+      <section class="user_info">
+        <%= render 'shared/user_info' %>
```
Comment (saobv-2223) [inline app/views/static_pages/home.html.erb]: using "

File: app/views/static_pages/home.html.erb [hunk 6]
```diff
@@ -1,13 +1,27 @@
...
+      </section>
+      <section class="micropost_form">
+        <%= render 'shared/micropost_form' %>
```
Comment (saobv-2223) [inline app/views/static_pages/home.html.erb]: using "

File: app/views/static_pages/home.html.erb [hunk 7]
```diff
@@ -1,13 +1,27 @@
...
+      </section>
+    </aside>
+    <div class="col-md-8">
+      <h3>Micropost Feed</h3>
+      <%= render 'shared/feed' %>
```
Comment (saobv-2223) [inline app/views/static_pages/home.html.erb]: using "

File: app/views/static_pages/home.html.erb [hunk 8]
```diff
@@ -1,13 +1,27 @@
...
+  </div>
+<% else %>
+  <div class="center jumbotron">
+    <h1><%= t ".title" %></h1>
+  <h2>
+    <%= t("home.page") %>
+    <%= link_to t("home.message"), Settings.links.rails_tutorial.home_page %>
+    <%= t("home.sample") %>
+  </h2>
+  <%= link_to t("signup"), signup_path, class: "btn btn-lg btn-primary" %>
```
Comment (saobv-2223) [inline app/views/static_pages/home.html.erb]:
```suggestion
    <h2>
      <%= t("home.page") %>
      <%= link_to t("home.message"), Settings.links.rails_tutorial.home_page %>
      <%= t("home.sample") %>
    </h2>
    <%= link_to t("signup"), signup_path, class: "btn btn-lg btn-primary" %>
```

**Comment by heimrih** [comment]
ready

**Comment by vanvtt-0952** [comment]
merged

**Pull request merged**